emf/loadflow_tool/model_merger/merge_functions.py

```diff
@@ -320,6 +322,97 @@ def remove_small_islands(solved_data, island_size_limit):
     return solved_data
 
 
+def get_opdm_data_from_models(model_data: list | pandas.DataFrame):
+    """
+    Check if input is already parsed to triplets. Do it otherwise
+    :param model_data: input models
+    :return triplets
+    """
+    if not isinstance(model_data, pandas.DataFrame):
+        model_data = load_opdm_data(model_data)
+    return model_data
+
+
+def get_boundary_nodes_between_igms(model_data: list | pandas.DataFrame):
+    """
+    Filters out nodes that are between the igms (mentioned at least 2 igms)
+    :param model_data: input models
+    : return series of node ids
+    """
+    model_data = get_opdm_data_from_models(model_data=model_data)
+    all_boundary_nodes = model_data[(model_data['KEY'] == 'TopologicalNode.boundaryPoint') &
+                                    (model_data['VALUE'] == 'true')]
+    # Get boundary nodes that exist in igms
+    nodes_in_igms = model_data[(model_data['KEY'] == 'SvVoltage.TopologicalNode') &
+                              (model_data['VALUE'].isin(all_boundary_nodes['ID']))]
+    # Get only those boundary nodes that exist in multiple igms
+    multiple_occurrences = nodes_in_igms['VALUE'].value_counts()
+    in_several_igms = multiple_occurrences[multiple_occurrences > 1]
+    return in_several_igms
+
+
+def take_best_match_for_sv_voltage(input_data, column_name: str = 'v'):
+    """
+    Returns one row for with sv voltage id for topological node
+    1) Take the first
+    2) If first is zero take first non-zero row if exists
+    :param input_data: input dataframe
+    :param column_name: name of the column
+    """
+    first_row = input_data.iloc[0]
+    remaining_rows = input_data[input_data[column_name] != 0]
+    if first_row['v'] == 0 and not remaining_rows.empty:
+        first_row = remaining_rows.iloc[0]
+    return first_row
+
+
+def copy_topological_nodes_over(cgm_sv_data, original_data: list | pandas.DataFrame):
+    """
+    If boundary point is between two igms that are merged then there might be a case when pypowsybl doesn't count it
+    In newer versions pypowsybl may count it from several sides (from each igm side). Therefore this copies sv
+    voltages from igms over to merged model
+    NOTE THAT THIS IS A HACK
+    :param cgm_sv_data: merged model as a dataframe
+    :param original_data: input models in OPDE format (dictionary)
+    :return updated merged model
+    """
+    some_data = get_opdm_data_from_models(model_data=original_data)
+    in_several_igms = get_boundary_nodes_between_igms(model_data=some_data)
+    # Get boundary nodes that already have the voltage level declared in sv file
+    already_existing = cgm_sv_data[(cgm_sv_data['KEY'] == 'SvVoltage.TopologicalNode') &
+                                   (cgm_sv_data['VALUE'].isin(in_several_igms.index))]
+    # Exclude the existing nodes
+    in_several_igms = in_several_igms[~in_several_igms.index.isin(already_existing['VALUE'])]
```

Comment:
Please do not use isin, it is known to be slow, there is function in triplet, filter data frame by data frame, if needed